### ISSUE TITLE
O1-1. DataService rather than DataNode saves binlog paths into Etcd

### ISSUE THREAD

**Issue by XuanYang-cn** [milvus-io/milvus#5220]
- [x] [design] 1. Add a new RPC function in dataservice, update data_service.proto.

- [x] [design] 2. Mv save-binlog-paths module from DataNode into DataService.

**Comment by XuanYang-cn** [comment]
**O2-3** DataNode is no longer aware of whether if segment flushed, so SegmentFlushed messages should be sent by DataService

**Comment by XuanYang-cn** [comment]
> 1. Add a new RPC function in dataservice, update data_service.proto. (Need to design)
> 2. Mv save-binlog-paths module from DataNode into DataService.

@sunby 
### 1. Proto design
```proto
rpc SaveBinlogPaths(SaveBinlogPathsRequest) returns (common.Status){}

// segmentID, fieldIDs, seg_binlog_paths are binlog paths for a segment.
// collectionID, ts_binlog_path and ddl_binlog_path are binlog paths for ddl message.
// start_positions and end_positions: msgstreams' message positions of datanode while flushing
message SaveBinlogPathsRequest {
   common.MsgBase base = 1;
   int64 segmentID = 2;
   int64 collectionID = 3;                                                   
   string ts_binlog_path = 4;                                                                                                                 
   string ddl_binlog_path = 5;
   repeated int64 fieldIDs = 6;
   repeated string seg_binlog_paths = 7;
   repeated internal.MsgPosition start_positions = 8;
   repeated internal.MsgPosition end_positions = 9;
 }
```
### 2. Save binlog paths in DataService

**Comment by XuanYang-cn** [comment]
@sunby @neza2017 After being updated about what's `autoflush` and what's `manul flush`, **Proto design NO.1** is the winner. But it needs some retouch and design docs should update the concepts of `auto-flush` and `manul-flush`

**Comment by XuanYang-cn** [comment]
### The Winner: proto design NO.1

```proto
rpc SaveBinlogPaths(SaveBinlogPathsRequest) returns (common.Status){}


message ID2PathList {                                                                                            
    int64 ID = 1;                                                                                                
    repeated string Paths = 2;                                                                                   
}                                                                                                                

message SaveBinlogPathsRequest {                                                                                 
    common.MsgBase base = 1; 
    int64 segmentID = 2;
    int64 collectionID = 3;
    ID2PathList field2BinlogPaths = 4;
    ID2PathList coll2TsBinlogPaths = 5; 
    ID2PathList coll2DdlBinlogPaths = 6;  
    repeated internal.MsgPosition start_positions = 7;  
    repeated internal.MsgPosition end_positions = 8; 
 }

```

**Comment by XuanYang-cn** [comment]
@sunby @congqixia I'll update the design docs

**Comment by XuanYang-cn** [comment]
- [ ] DDL flushing procedure needs to redesign, but not in high priority.

**Comment by XuanYang-cn** [comment]
> **O2-3** DataNode is no longer aware of whether if segment flushed, so SegmentFlushed messages should be sent by DataService[after O1-1]
> 
> 1. DataNode removes this code, doesn't need a flush-completed channel.
> 2. DataService publishes flush-completed message into its channel.

Could you do the DataService part of this task? @congqixia

**Comment by XuanYang-cn** [comment]
> * [ ]  DDL flushing procedure needs to redesign, but not in high priority.

Now blocked by **O2-3**, will do this task next.

**Comment by XuanYang-cn** [comment]
> > * [x]   DDL flushing procedure needs to redesign, but not in high priority.
> 
> Now blocked by **O2-3**, will do this task next. #5289

Now waiting for #5277

The final PR for O1-1 is migrating the current save-etcd and completed-message-publish procedure into DataService

```
# etcd paths for ddl binlog paths
${prefix}/${collectionID}/${segmentID}/${idx}
```

```
message SaveBinlogPathsRequest {
    common.MsgBase base = 1;
    int64 segmentID = 2;
    int64 collectionID = 3;
    ID2PathList field2BinlogPaths = 4;
    repeated DDLBinlogMeta = 5;
    repeated internal.MsgPosition start_positions = 7;  
    repeated internal.MsgPosition end_positions = 8; 
 }
```

**Comment by XuanYang-cn** [comment]
1. Correct flush procedure :100: 
- design docs :ok: 
- Insert log auto-flush will buffer binlog paths :ok:
- DDL buffer doesn't flush automatically :ok:
2. DataNode doesn't interact with Etcd
- design docs :ok:
- DataService save binlog paths @congqixia :ok:
- DataService publish flush completed message @congqixia :ok:
- DataNode and DataService complete flush procedure and pass CI
- Remove outdated procedure in DataNode
3. DataNode Service Registration & Discovery @godchen0212 
4. DataNod Recovery
- design docs
- Set positions in DDL message pack @xige-16 
- DatatNode Scales flowgraph
- DataNode handles msgpositions
- Idempotent of `WatchDmChannels`, `FlushSegment`

**Comment by sunby** [comment]
`SaveBinlogPaths` rpc maybe fail because of crash of DataService. DataNodes should keep the content of this request in memory until receiving a successful response.  If DataNode crashes during the process,  it will seek to the correct position according to the position recorded in DataService after it restarts. If DataService did not receive or correctly handle the request, it will ask DataNode to restore the last position before sending the `SaveBinlogPaths` request. If DataService correctly handle the request, then DataNode should seek to the next position. So no segment will be lost.

**Comment by sunby** [comment]
In DataService, these are the process steps of `SaveBinlogPaths` request: 
1. Transactionally save binlog and position meta& update segment state to `Flushing`.
2. Send `SegmentFlushMsg` to msg stream 
3. Update segment state to `Flushed`
Once segment state is set to `Flushing` and server crashes afterwards. DataService will retry to flush the segment after it restarts.
If step 2 fails and server is alive, it will return unsuccessful response. DataService will reflush the segment.

**Comment by XuanYang-cn** [comment]
> In DataService, these are the process steps of `SaveBinlogPaths` request:
> 
> 1. Transactionally save binlog and position meta& update segment state to `Flushing`.
> 2. Send `SegmentFlushMsg` to msg stream
> 3. Update segment state to `Flushed`
>    Once segment state is set to `Flushing` and server crashes afterwards. DataService will retry to flush the segment after it restarts.
>    If step 2 fails and server is alive, it will return unsuccessful response. DataService will reflush the segment.

Thanks @sunby , I'll give my reply soon, currently unavailable.

**Comment by XuanYang-cn** [comment]
> In DataService, these are the process steps of `SaveBinlogPaths` request:
> 
> 1. Transactionally save binlog and position meta& update segment state to `Flushing`.
> 2. Send `SegmentFlushMsg` to msg stream
> 3. Update segment state to `Flushed`
>    Once segment state is set to `Flushing` and server crashes afterwards. DataService will retry to flush the segment after it restarts.
>    If step 2 fails and server is alive, it will return unsuccessful response. DataService will reflush the segment.

@sunby Great! A little touch up:

1. Transactionally save binlog and position meta && update segment state to 'Flushed' **&& save `SegmentFlushMsg` into etcd**
2. Pulish `SegmentFlushMsg` into msgstream
3. Delete `SegmentFlushMsg` in Etcd
  - If step 2 fails and server is alive, server redos step 2&3
  - If server crashes between step 1 and step2, when server initializes, and sees a `Flushed` segment with `SegmentFlushMsg` in etcd, it redos step 2&3.
  - If step 3 fails, just ignore. Because MasterService handles `SegmentFlushMsg` in  idempotent.

The main difference of this touching up is that saving binlog and position meta into etcd is a sign of **a successful flush**. 

We don't want to waste DataNode resources for things can be redo by DataService.

**Comment by XuanYang-cn** [comment]
Close for already done
